internal/controller/customreplicaset_controller.go

```diff
@@ -332,215 +337,219 @@ func (r *CustomReplicaSetReconciler) findChildPods(ctx context.Context, req ctrl
+	} else if totalAvailPods > int(crs.Spec.Replicas) {
+		podsToDelete := totalAvailPods - int(crs.Spec.Replicas)
+		if err := r.deletePods(ctx, crs, podsToDelete, podsPerRevision, childPods, false); err != nil {
+			fmt.Println("Failed to delete pods", err)
 			return err
 		}
 	} else {
-		if err := r.upgradeOrDowngradePods(ctx, availableStdPods, availableUpgPods, crs, childPods, log); err != nil {
-			log.Error(err, "Failed to upgrade or downgrade pods")
+		if err := r.checkUpgradePods(ctx, crs, totalAvailPods, podsPerRevision, latestRevision, childPods); err != nil {
+			fmt.Println("Failed to upgrade or downgrade pods", err)
 			return err
 		}
 	}
 	return nil
 }
 
-func (r *CustomReplicaSetReconciler) createPods(ctx context.Context, totalPods, stdPods, upgPods int, cr *customreplicasetv1.CustomReplicaSet, log logr.Logger) error {
-	for totalPods < int(cr.Spec.Replicas) {
-		var newPod *corev1.Pod
-		// Create any missing upgraded pods first
-		if upgPods < int(cr.Spec.Partition) {
-			newPod = r.newPodForCR(cr, true)
-			upgPods++
-		} else {
-			newPod = r.newPodForCR(cr, false)
-			stdPods++
+func (r *CustomReplicaSetReconciler) createPods(ctx context.Context, cr *customreplicasetv1.CustomReplicaSet, podsToCreate int, podsPerRevision map[int][]*corev1.Pod, revision *v1.ControllerRevision) error {
+	for podsToCreate > 0 {
+		// If we are short pods, create new pods of latest revision
+		newPod, err := r.newPodForCR(cr, revision)
+		if err != nil {
+			fmt.Println("Unable to initialize new pod for creation", err)
+			return err
 		}
 
 		if err := r.Create(ctx, newPod); err != nil {
-			log.Error(err, "Unable to create new pod")
+			fmt.Println("Unable to create new pod", err)
 			return err
 		}
 		fmt.Println("Created Pod")
-		totalPods++
+		podsToCreate--
 	}
 	return nil
 }
 
-func (r *CustomReplicaSetReconciler) deletePods(ctx context.Context, totalPods, stdPods, upgPods int, cr *customreplicasetv1.CustomReplicaSet, childPods corev1.PodList, log logr.Logger) error {
-	podsToDelete := totalPods - int(cr.Spec.Replicas)
-	upgPodsToDelete := upgPods - int(cr.Spec.Partition)
-	stdPodsToDelete := podsToDelete - upgPodsToDelete
+func (r *CustomReplicaSetReconciler) deletePods(ctx context.Context, cr *customreplicasetv1.CustomReplicaSet, podsToDelete int, podsPerRevision map[int][]*corev1.Pod, childPods corev1.PodList, reverse bool) error {
+	sortedKeys := sortKeys(podsPerRevision, reverse)
+	finishedDeletion := false
 
-	for _, pod := range childPods.Items {
-		if pod.Status.Phase == corev1.PodRunning || pod.Status.Phase == corev1.PodPending {
-			if strings.Contains(pod.ObjectMeta.Name, "upgraded") {
-				if upgPodsToDelete > 0 {
-					upgPodsToDelete--
-				} else {
+	for _, key := range sortedKeys {
+		if finishedDeletion {
+			return nil
+		}
+		for _, pod := range podsPerRevision[key] {
+			if pod.Status.Phase == corev1.PodRunning || pod.Status.Phase == corev1.PodPending {
+				if pod.DeletionTimestamp != nil {
```

Comment:
Why do you want to check this? 

Delete a pod with DeletionTimestamp set does not hurt, right? The K8s REST APIs are idempotent.

Reply:
This is like a check that if a node is already marked for deletion...we don't need to re-delete it...saves a delete operation

Reply:
Followed up offline! Removed this check